doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,297 @@
+
+**Maintainable** because relevancy tuning can be automated to not become
+stale.
+
+**Successful** because results should achieve higher nDCG scores and
+click-through rates than we can achieve through manual tuning alone.
+
+### The problem
+
+Manual relevancy tuning can work, but doesn't scale well. Eventually
+the process becomes a game of whack-a-mole. If we downgrade popularity
+boosting, then this query might have better results, but other queries
+might have worse results.
+
+Hand-tuning relevance doesn't work very well because field boosts
+are complex, and difficult to maintain.
+
+Moreover field boosts might have dependencies: some document types might
+benefit from a stronger title boost than other document types, for
+instance. Expressing that just with Elasticsearch would be prohibitively
+complex, and be almost impossible to maintain.
+
+### Validating Learning To Rank
+
+We assumed that we would be able to get better results using Learning
+To Rank. To validate this assumption, we worked on a spike.
+
+In a [validation spike][] we found that we could build a model that provided
+a better nDCG score than our current search algorithm for the queries we
+have manually gathered relevancy judgements for.
+
+This validated for us that Learning To Rank is feasible and could
+lead to more relevant results for users.
+
+Though Learning To Rank won't make relevance tuning easier, it could
+make search results more relevant.
+
+### Final decision
+
+We feel confident about implementing Learning To Rank (LTR) for the
+following reasons:
+
+- We think that it will improve search relevancy for citizens, enabling
+  them to find what they need on GOV.UK
+- We have validated our assumptions with a spike
+- LTR is a well trodden path, and other organisations have written about
+  their implementations with varying levels of openness (e.g. Wikimedia)
+- GOV.UK has implemented other machine learning projects with success
+  (related links, topic taxonomy), and we have a data science team which
+  the search team has already worked with
+- TensorFlow provides an off-the-shelf ranking module, which solves
+  some of the machine learning problems of LTR for us
+- Manual relevancy tuning is not scalable, and won't get us to the same
+  level of relevancy as LTR
+
+## Implementation plan
+
+We have decided to implement Learning To Rank by following on from
+the work started in the [validation spike][].
+
+As we are at the beginning of the project, this section may not
+accurately reflect the final implementation.
+
+[validation spike]: https://github.com/alphagov/search-api/pull/1768
+
+### Machine learning library
+
+We have decided to use [TensorFlow's ranking library][] for our
+implementation of Learning To Rank.
+
+TensorFlow is well supported, and the ranking library is under active development.
+
+We considered using the [elasticsearch-learning-to-rank][] plugin.
+However, AWS Elasticsearch does not permit us to install custom plugins.
```

Comment:
That's a shame! Is using another hosted elasticsearch solution an option?

Reply:
I think it's unlikely, since we're migrating most things into AWS at the moment. But it's worth asking the question - I'll see what our tech archs think :)

Reply:
No, not another ElasticSearch migration!